### PR TITLE
Add function attributes to JSON ABI

### LINKED ISSUE #446: Add function attributes to JSON ABI
This is needed to support the `#[payable]` annotation in the Fuel SDKs. See the PR introducing the payability analysis to Sway: https://github.com/FuelLabs/sway/pull/3450.

### PR DESCRIPTION
**Pull request by anton-trunov**
close #446

I went for one of the simplest solutions here. I'd propose to switch to a more complex attribute representation when the need arises -- the current format does not support attributes with parameters. But I'm happy to take suggestions on this as well and switch to a more general schema if someone has potential use cases in mind.

### TODOs based on an internal discussion

- [x] Document possible function attributes
- [x] Add doc comment attributes
- [x] Add storage attributes

### CODE REVIEW

**Review by mohammadfawaz — APPROVED** [review]
LGTM! Thanks.

File: src/protocol/abi/json_abi_format.md [hunk 1]
```diff
@@ -49,6 +49,11 @@ The ABI of a contract is represented as a JSON object containing the following p
     - `"typeArguments"`: an array of the _type arguments_ used when applying the type of the output, if the type is generic, and `null` otherwise. Each _type argument_ is a _type application_ represented as a JSON object that contains the following properties:
       - `"type"`: the _type declaration_ ID of the type of the _type argument_.
       - `"typeArguments"`: an array of the _type arguments_ used when applying the type of the _type argument_, if the type is generic, and `null` otherwise. The format of the elements of this array recursively follows the rules described in this section.
+  - `"attributes"`: an optional array of _attributes_. Each _attribute_ is represented as a JSON object that contains
+  the following properties:
+    - `"name"`: the name of the attribute (the only valid attribute names are `storage`, `payable`, `test`, `inline`,
```
Comment (mohammadfawaz) [inline src/protocol/abi/json_abi_format.md:54]: I propose that we make this sentence less specific about what attributes are supported, and move the actual supported attributes into a separate section in this file where each attribute is listed with exactly what it means and what it implies (e.g. `payable` implies that no coins can be forwarded when calling this method, etc.). We can point to that section here by saying: "For a list of supported attributes, see ..."

Comment (anton-trunov) [reply src/protocol/abi/json_abi_format.md]: Great suggestion -- I took a stab at it

File: src/protocol/abi/json_abi_format.md [hunk 2]
```diff
@@ -74,6 +77,17 @@ The ABI of a contract is represented as a JSON object containing the following p
 
 > **Note**: This JSON should be both human-readable and parsable by the tooling around the FuelVM and the Sway programming language. There is a detailed specification for the binary encoding backing this readable descriptor. The [Function Selector Encoding](./fn_selector_encoding.md) section specifies the encoding for the function being selected to be executed and each of the argument types.
 
+### Attributes Semantics
+
+| Attribute name | Attribute arguments               | Semantics |
```
Comment (mohammadfawaz) [inline src/protocol/abi/json_abi_format.md:82]: Do you mind "prettyfying" this table? I use [this](https://marketplace.visualstudio.com/items?itemName=darkriszty.markdown-table-prettify) VS code extension for that purpose.

Comment (anton-trunov) [reply src/protocol/abi/json_abi_format.md]: Sure. Just did that

**Review by mohammadfawaz — APPROVED** [review]
Nice.

**Pull request merged**